### PR TITLE
Add Hoodoo::TransientStore

### PR DESCRIPTION
**Pull request by pond**
Long-awaited. Addresses part of https://github.com/LoyaltyNZ/hoodoo/issues/105, but the session engine still uses Memcached directly.

### CODE REVIEW

File: lib/hoodoo/transient_store/transient_store/memcached_redis_mirror.rb [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      # Permitted values are:
+      #
+      # +:memcached+:: Only Memcached needs to have a value for a key
+      # +:redis+::     Only Redis needs to have a value for a key
+      # +both+::       Both engines must have the key
+      #
+      # This is useful in migration scenarios where moving from Memcached to
+      # Redis or vice versa. If wishing to be able to still read old data only
+      # in Memcached, set +:memcached+; else +:redis+. For true mirroring,
+      # insist on both with +:both+.
```
Comment (aames) [inline lib/hoodoo/transient_store/transient_store/memcached_redis_mirror.rb]: Expanding upon this paragraph by providing clear example instructions for either migration scenario would be really helpful.

File: spec/transient_store/transient_store/memcached_redis_mirror_spec.rb [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        expect_any_instance_of( Hoodoo::TransientStore::Redis     ).to receive( :set ).with( key: @key, payload: @payload, maximum_lifespan: @ttl ).and_call_original()
+
+        @instance.set(
+          key:              @key,
+          payload:          @payload,
+          maximum_lifespan: @ttl
+        )
+      end
+
+      it 'deletes known keys' do
```
Comment (aames) [inline spec/transient_store/transient_store/memcached_redis_mirror_spec.rb]: delete successful when key exists only in redis or memcached would be good here

File: spec/transient_store/transient_store/memcached_redis_mirror_spec.rb [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+           after( :each ) { @instance.get_keys_from = :both  }
+
+          it 'returns data if the key is in both engines' do
+            expect_any_instance_of( Hoodoo::TransientStore::Memcached ).to     receive( :get ).with( key: @key ).and_call_original()
+            expect_any_instance_of( Hoodoo::TransientStore::Redis     ).to_not receive( :get )
+
+            expect( @instance.get( key: @key ) ).to eql( @payload )
+          end
+
+          it 'returns data if the key is only in Memcached' do
```
Comment (aames) [inline spec/transient_store/transient_store/memcached_redis_mirror_spec.rb]: It would be interesting to also have a GET test for when the key exists in both but the value is different.

**Comment by aames** [comment]
Fantastic - I could only make a couple of suggestions for edge cases in test coverage and a request for expanding upon an example.

File: spec/transient_store/transient_store/memcached_redis_mirror_spec.rb (hunk 3 above)
Comment (pond) [inline spec/transient_store/transient_store/memcached_redis_mirror_spec.rb]: IRL discussion via Slack led us to agree that the performance impact makes it not worth addressing this case in code, but I'll add RDoc coverage indicating that such a synchronisation failure or edge case deliberate-manual-write-to-one-of-the-engines issue will lead to (Redis-or-Memcached)'s value being chosen in preference.

File: lib/hoodoo/transient_store/transient_store/memcached_redis_mirror.rb (hunk 1 above)
Comment (pond) [inline lib/hoodoo/transient_store/transient_store/memcached_redis_mirror.rb]: Addressed.

File: spec/transient_store/transient_store/memcached_redis_mirror_spec.rb (hunk 3 above)
Comment (pond) [inline spec/transient_store/transient_store/memcached_redis_mirror_spec.rb]: (Done)

File: spec/transient_store/transient_store/memcached_redis_mirror_spec.rb (hunk 2 above)
Comment (pond) [inline spec/transient_store/transient_store/memcached_redis_mirror_spec.rb]: Good catch, added.
